File: src/print_partner/core/plate_packer.py
"""Bin-pack part meshes onto printer beds (multi-plate)."""

from __future__ import annotations

from dataclasses import dataclass
from copy import deepcopy

import trimesh

from print_partner.core.filament_assigner import PartCopy
from print_partner.core.printer_fleet import PrinterMachine


@dataclass
class PlacedItem:
    copy: PartCopy
    mesh: trimesh.Trimesh
    x_mm: float
    y_mm: float
    width_mm: float
    depth_mm: float
    height_mm: float


@dataclass
class PlateLayout:
    printer_id: str
    index: int
    items: list[PlacedItem]


def load_mesh_for_copy(copy: PartCopy) -> tuple[trimesh.Trimesh | None, str | None]:
    part = copy.part
    stl_path = part.absolute_path
    if not stl_path or not stl_path.is_file():
        return None, f"Missing STL: {part.relative_path}"
    try:
        loaded = trimesh.load(stl_path, force="mesh")
    except Exception as exc:
        return None, f"Could not load {part.relative_path}: {exc}"

    if isinstance(loaded, trimesh.Scene):
        meshes = [g for g in loaded.geometry.values() if isinstance(g, trimesh.Trimesh)]
        mesh = trimesh.util.concatenate(meshes) if meshes else None
    elif isinstance(loaded, trimesh.Trimesh):
        mesh = loaded
    else:
        mesh = None

    if mesh is None or mesh.vertices.shape[0] == 0:
        return None, f"Empty mesh: {part.relative_path}"
    return mesh, None


def _place_on_bed(mesh: trimesh.Trimesh, x: float, y: float) -> trimesh.Trimesh:
    m = deepcopy(mesh)
    bounds = m.bounds
    m.apply_translation([x - bounds[0][0], y - bounds[0][1], -bounds[0][2]])
    return m


def pack_copies_on_printer(
    printer: PrinterMachine,
    copies: list[PartCopy],
    *,
    spacing_mm: float | None = None,
) -> tuple[list[PlateLayout], list[str]]:
    """Shelf-pack copies onto one or more plates for a single printer."""
    warnings: list[str] = []
    if not copies:
        return [], warnings

    margin = printer.margin_mm
    spacing = spacing_mm if spacing_mm is not None else margin
    bed_w = printer.bed_width_mm - 2 * margin
    bed_d = printer.bed_depth_mm - 2 * margin
    max_z = printer.bed_height_mm

    loaded: list[tuple[PartCopy, trimesh.Trimesh, float, float, float]] = []
    for copy in copies:
        mesh, err = load_mesh_for_copy(copy)
        if err:
            warnings.append(err)
            continue
        assert mesh is not None
        w, d, h = (float(x) for x in mesh.extents)
        if w > bed_w or d > bed_d:
            warnings.append(
                f"{copy.part.filename} ({w:.0f}×{d:.0f} mm) too large for "
                f"{printer.name} bed ({printer.bed_width_mm:.0f}×{printer.bed_depth_mm:.0f} mm)"
            )
            continue
        if max_z is not None and h > max_z:
            warnings.append(
                f"{copy.part.filename} height {h:.0f} mm exceeds "
                f"{printer.name} Z limit {max_z:.0f} mm"
            )
        loaded.append((copy, mesh, w, d, h))

    loaded.sort(key=lambda t: max(t[2], t[3]), reverse=True)

    plates: list[PlateLayout] = []
    current_items: list[PlacedItem] = []
    layout_x = 0.0
    layout_y = 0.0
    row_height = 0.0
    plate_index = 1

    def flush_plate() -> None:
        nonlocal current_items, layout_x, layout_y, row_height, plate_index
        if current_items:
            plates.append(
                PlateLayout(printer_id=printer.id, index=plate_index, items=current_items)
            )
            plate_index += 1
        current_items = []
        layout_x = 0.0
        layout_y = 0.0
        row_height = 0.0

    for copy, mesh, width, depth, height in loaded:
        if layout_x > 0 and layout_x + width > bed_w:
            layout_x = 0.0
            layout_y += row_height + spacing
            row_height = 0.0

        if layout_y + depth > bed_d:
            flush_plate()

        placed_mesh = _place_on_bed(mesh, margin + layout_x, margin + layout_y)
        current_items.append(
            PlacedItem(
                copy=copy,
                mesh=placed_mesh,
                x_mm=margin + layout_x,
                y_mm=margin + layout_y,
                width_mm=width,
                depth_mm=depth,
                height_mm=height,
            )
        )
        layout_x += width + spacing
        row_height = max(row_height, depth)

    flush_plate()
    return plates, warnings


def pack_single_plate(
    printer: PrinterMachine,
    copies: list[PartCopy],
    *,
    plate_index: int = 1,
    spacing_mm: float | None = None,
) -> tuple[PlateLayout | None, list[str]]:
    """Pack copies onto one logical plate (warn if auto-pack would need more beds)."""
    if not copies:
        return None, []
    plates, warnings = pack_copies_on_printer(printer, copies, spacing_mm=spacing_mm)
    if not plates:
        return None, warnings
    if len(plates) > 1:
        warnings = list(warnings) + [
            f"Plate {plate_index} on {printer.name}: parts may not fit one bed "
            f"(auto-pack used {len(plates)} plates)."
        ]
    items: list[PlacedItem] = []
    for plate in plates:
        items.extend(plate.items)
    return PlateLayout(printer_id=printer.id, index=plate_index, items=items), warnings


def estimate_plate_counts(
    printers: list[PrinterMachine],
    by_printer: dict[str, list[PartCopy]],
) -> dict[str, int]:
    counts: dict[str, int] = {}
    for printer in printers:
        copies = by_printer.get(printer.id, [])
        if not copies:
            counts[printer.id] = 0
            continue
        plates, _ = pack_copies_on_printer(printer, copies)
        counts[printer.id] = len(plates)
    return counts

File: web/packages/domain/src/plate-packer.ts
import type { PartCopy } from "./checkoff-missing.js";
import { folderKeyFromRelativePath } from "./parts-grouping.js";
import { repoNameFromSourceLayer } from "./parts-tree.js";
import type { MergePartExport, PrinterMachine } from "./filament-assigner.js";
import { loadStlMesh, placeMeshOnBed, type StlMesh } from "./stl-mesh.js";

export type HeightBand = "flat" | "short" | "medium" | "tall" | "very-tall";

/**
 * Height band thresholds (mm), chosen to reflect common 3D-printing part
 * height use cases:
 *   flat:      < 10 mm   — baseplates, badges, thin panels
 *   short:     10–50 mm  — brackets, small enclosures, spacers
 *   medium:    50–150 mm — most functional parts, boxes, housings
 *   tall:      150–300 mm — vases, tall enclosures, most single-plate maximum
 *   very-tall: > 300 mm  — full-height prints near/above typical printer Z limits
 * Upper bound of each band is exclusive; the next band's lower bound is inclusive,
 * i.e. a part exactly at a threshold value falls into the taller band.
 */
export const HEIGHT_BAND_THRESHOLDS_MM = {
  flat: 10,
  short: 50,
  medium: 150,
  tall: 300,
} as const;

/**
 * Classify a part height (mm) into one of five height bands.
 * Pure function: given the same input it always returns the same output.
 *
 * Negative heights are invalid mesh data; they are classified as "flat" (the
 * lowest band) rather than throwing, so callers can surface a separate
 * validation warning without this classification itself failing. NaN is
 * likewise treated as "flat" (safest default for unknown/bad data).
 */
export function classifyHeightBand(heightMm: number): HeightBand {
  if (Number.isNaN(heightMm) || heightMm < HEIGHT_BAND_THRESHOLDS_MM.flat) return "flat";
  if (heightMm < HEIGHT_BAND_THRESHOLDS_MM.short) return "short";
  if (heightMm < HEIGHT_BAND_THRESHOLDS_MM.medium) return "medium";
  if (heightMm < HEIGHT_BAND_THRESHOLDS_MM.tall) return "tall";
  return "very-tall";
}

/** Human-readable labels for each height band, used as plate `group_label` when packing by Height Band. */
export const HEIGHT_BAND_LABELS: Record<HeightBand, string> = {
  flat: "Flat (<10mm)",
  short: "Short (10–50mm)",
  medium: "Medium (50–150mm)",
  tall: "Tall (150–300mm)",
  "very-tall": "Very Tall (>300mm)",
};

/** Iteration order from shortest to tallest band, used when grouping/sorting by Height Band. */
export const HEIGHT_BAND_ORDER: HeightBand[] = ["flat", "short", "medium", "tall", "very-tall"];

/**
 * Strategy used to group parts into plates.
 *   location:    group by filament + source repo/folder (existing default behavior)
 *   height_band: group by classifyHeightBand(bounds.heightMm) — see packCopiesGroupedByHeightBand
 */
export type GroupingStrategy = "location" | "height_band";

function partFilamentLabel(part: MergePartExport): string {
  const label = (part.filamentDisplay ?? part.filament_display ?? "").trim();
  if (label) return label;
  const colorId = part.filamentColorId ?? part.filament_color_id;
  if (colorId) return colorId;
  if (part.role) return `(filament not set — ${part.role})`;
  return "(filament not set)";
}

export type PlacedItem = {
  copy: PartCopy;
  mesh: StlMesh;
  x_mm: number;
  y_mm: number;
  width_mm: number;
  depth_mm: number;
  height_mm: number;
  heightBand: HeightBand;
};

export type PlateLayout = {
  printer_id: string;
  index: number;
  items: PlacedItem[];
  group_label: string;
};

function loadMeshForCopy(copy: PartCopy): [StlMesh | null, string | null] {
  const part = copy.part as MergePartExport;
  const stlPath = part.absolutePath;
  if (!stlPath) return [null, `Missing STL: ${part.relativePath}`];
  try {
    const mesh = loadStlMesh(stlPath);
    if (!mesh) return [null, `Could not load ${part.relativePath}`];
    return [mesh, null];
  } catch (e) {
    return [null, `Could not load ${part.relativePath}: ${e instanceof Error ? e.message : String(e)}`];
  }
}

/**
 * Check a completed plate's height variance and warn when the spread between
 * the tallest and shortest part exceeds 2x the shortest part's height. Applies
 * uniformly to every plate produced by packCopiesOnPrinter, regardless of which
 * grouping strategy (location, height band, or none) selected the parts on it —
 * mixed-height plates cause uneven cooling / support needs during printing.
 *
 * Returned so callers can surface it to the user (e.g. as a toast) alongside
 * other pack warnings — see
 * export3mfJobResult.ts / exportStlJobResult.ts for the existing
 * warnings-array-to-toast pattern this plugs into.
 */
function checkPlateHeightVariance(
  printer: PrinterMachine,
  plateIndex: number,
  items: PlacedItem[],
): string | null {
  const heights = items.map((i) => i.height_mm).filter((h) => Number.isFinite(h) && h > 0);
  if (heights.length < 2) return null;
  const minH = Math.min(...heights);
  const maxH = Math.max(...heights);
  const variance = maxH - minH;
  if (minH <= 0 || variance <= 2 * minH) return null;
  const warning =
    `Plate ${plateIndex} on ${printer.name}: height variance ${variance.toFixed(1)} mm exceeds ` +
    `2× the shortest part (${minH.toFixed(1)} mm) — consider grouping by Height Band.`;
  return warning;
}

export function packCopiesOnPrinter(
  printer: PrinterMachine,
  copies: PartCopy[],
  options?: { spacing_mm?: number | null },
): [PlateLayout[], string[]] {
  const warnings: string[] = [];
  if (!copies.length) return [[], warnings];

  const margin = printer.margin_mm;
  const spacing = options?.spacing_mm ?? margin;
  const bedW = printer.bed_width_mm - 2 * margin;
  const bedD = printer.bed_depth_mm - 2 * margin;
  const maxZ = printer.bed_height_mm;

  const loaded: Array<[PartCopy, StlMesh, number, number, number]> = [];
  for (const copy of copies) {
    const [mesh, err] = loadMeshForCopy(copy);
    if (err) {
      warnings.push(err);
      continue;
    }
    if (!mesh) continue;
    const w = mesh.bounds.widthMm;
    const d = mesh.bounds.depthMm;
    const h = mesh.bounds.heightMm;
    if (w > bedW || d > bedD) {
      warnings.push(
        `${(copy.part as MergePartExport).filename} (${w.toFixed(0)}×${d.toFixed(0)} mm) too large for ` +
          `${printer.name} bed (${printer.bed_width_mm.toFixed(0)}×${printer.bed_depth_mm.toFixed(0)} mm)`,
      );
      continue;
    }
    if (maxZ != null && h > maxZ) {
      warnings.push(
        `${(copy.part as MergePartExport).filename} height ${h.toFixed(0)} mm exceeds ` +
          `${printer.name} Z limit ${maxZ.toFixed(0)} mm`,
      );
    }
    loaded.push([copy, mesh, w, d, h]);
  }

  loaded.sort((a, b) => Math.max(b[2], b[3]) - Math.max(a[2], a[3]));

  const totalFootprint = loaded.reduce((sum, [, , w, d]) => sum + w * d, 0);
  const bedArea = Math.max(bedW, 0) * Math.max(bedD, 0);
  if (bedArea > 0 && totalFootprint > bedArea * 0.9 && loaded.length > 1) {
    const estPlates = Math.max(2, Math.floor(totalFootprint / bedArea) + 1);
    warnings.push(
      `Estimated ${estPlates} plate(s) needed on ${printer.name} ` +
        `(${totalFootprint.toFixed(0)} mm² footprint on ${bedArea.toFixed(0)} mm² bed).`,
    );
  }

  const plates: PlateLayout[] = [];
  let currentItems: PlacedItem[] = [];
  let layoutX = 0;
  let layoutY = 0;
  let rowHeight = 0;
  let plateIndex = 1;

  const flushPlate = () => {
    if (currentItems.length) {
      const varianceWarning = checkPlateHeightVariance(printer, plateIndex, currentItems);
      if (varianceWarning) warnings.push(varianceWarning);
      plates.push({
        printer_id: printer.id,
        index: plateIndex,
        items: currentItems,
        group_label: "",
      });
      plateIndex += 1;
    }
    currentItems = [];
    layoutX = 0;
    layoutY = 0;
    rowHeight = 0;
  };

  for (const [copy, mesh, width, depth, height] of loaded) {
    if (layoutX > 0 && layoutX + width > bedW) {
      layoutX = 0;
      layoutY += rowHeight + spacing;
      rowHeight = 0;
    }
    if (layoutY + depth > bedD) flushPlate();

    const placedMesh = placeMeshOnBed(mesh, margin + layoutX, margin + layoutY);
    currentItems.push({
      copy,
      mesh: placedMesh,
      x_mm: margin + layoutX,
      y_mm: margin + layoutY,
      width_mm: width,
      depth_mm: depth,
      height_mm: height,
      heightBand: classifyHeightBand(height),
    });
    layoutX += width + spacing;
    rowHeight = Math.max(rowHeight, depth);
  }
  flushPlate();
  return [plates, warnings];
}

export function packCopiesGroupedByLocation(
  printer: PrinterMachine,
  copies: PartCopy[],
  options?: { spacing_mm?: number | null },
): [PlateLayout[], string[]] {
  if (!copies.length) return [[], []];
  const groups: Record<string, PartCopy[]> = {};
  for (const copy of copies) {
    const part = copy.part as MergePartExport;
    const key = [
      partFilamentLabel(part),
      repoNameFromSourceLayer(part.sourceLayer),
      folderKeyFromRelativePath(part.relativePath),
    ].join("\0");
    (groups[key] ??= []).push(copy);
  }

  const allPlates: PlateLayout[] = [];
  const allWarnings: string[] = [];
  let plateIndex = 1;
  const sortedKeys = Object.keys(groups).sort((a, b) => a.localeCompare(b));
  for (const key of sortedKeys) {
    const [filament, repo, folder] = key.split("\0");
    const folderDisp = folder === "(root)" ? "root" : folder;
    const label = `${filament} · ${repo} · ${folderDisp}`;
    const [plates, warnings] = packCopiesOnPrinter(printer, groups[key], options);
    for (const plate of plates) {
      plate.index = plateIndex;
      plate.group_label = label;
      plateIndex += 1;
      allPlates.push(plate);
    }
    allWarnings.push(...warnings);
  }
  return [allPlates, allWarnings];
}

/**
 * Group parts into their height bands (flat/short/medium/tall/very-tall) and
 * pack each band onto its own plate(s), ordered shortest band to tallest.
 * Used when the active grouping strategy is "Height Band" — see GroupingStrategy.
 *
 * Height is classified from the same mesh bounds.heightMm used by
 * packCopiesOnPrinter, via classifyHeightBand. A part whose STL fails to load
 * is skipped here (as in packCopiesOnPrinter) and its load error is still
 * surfaced through the returned warnings.
 */
export function packCopiesGroupedByHeightBand(
  printer: PrinterMachine,
  copies: PartCopy[],
  options?: { spacing_mm?: number | null },
): [PlateLayout[], string[]] {
  if (!copies.length) return [[], []];

  const bandGroups: Record<HeightBand, PartCopy[]> = {
    flat: [],
    short: [],
    medium: [],
    tall: [],
    "very-tall": [],
  };
  const allWarnings: string[] = [];

  for (const copy of copies) {
    const [mesh, err] = loadMeshForCopy(copy);
    if (err) {
      allWarnings.push(err);
      continue;
    }
    if (!mesh) continue;
    const band = classifyHeightBand(mesh.bounds.heightMm);
    bandGroups[band].push(copy);
  }

  const allPlates: PlateLayout[] = [];
  let plateIndex = 1;
  for (const band of HEIGHT_BAND_ORDER) {
    const bandCopies = bandGroups[band];
    if (!bandCopies.length) continue;
    // Sort tallest-first within the band so the shelf packer places the largest
    // footprints first (matches the sort packCopiesOnPrinter would otherwise do).
    const [plates, warnings] = packCopiesOnPrinter(printer, bandCopies, options);
    for (const plate of plates) {
      plate.index = plateIndex;
      plate.group_label = HEIGHT_BAND_LABELS[band];
      plateIndex += 1;
      allPlates.push(plate);
    }
    allWarnings.push(...warnings);
  }
  return [allPlates, allWarnings];
}

/**
 * Dispatch to the packer for the given grouping strategy. Height Band grouping
 * sorts/groups parts into their bands (flat → very-tall) before shelf-packing;
 * Location grouping (default) buckets by filament + source repo/folder.
 */
export function packCopiesGrouped(
  strategy: GroupingStrategy,
  printer: PrinterMachine,
  copies: PartCopy[],
  options?: { spacing_mm?: number | null },
): [PlateLayout[], string[]] {
  if (strategy === "height_band") {
    return packCopiesGroupedByHeightBand(printer, copies, options);
  }
  return packCopiesGroupedByLocation(printer, copies, options);
}
